ScalecubeConfigurationServiceConfigSource moved from scalecube-config

aharonha opened the pull request:
https://github.com/scalecube/scalecube-config/pull/108

Commit:
ScalecubeConfigurationServiceConfigSource moved from scalecube-config

Commit:
config source module

Commit:
the truth is that the value is and will always be JSON!!!

Commit:
no need for schema, it's a JSON object.

Commit:
fixtures require TestTemplate, not ParameterizedTest

segabriel commented:
Please look at my proposal about ConfigurationServiceConfigSource: https://github.com/scalecube/scalecube-configuration-service/pull/123

Commit:
token is string, data is json.

Commit:
removed unused imports

Commit:
revert changes made on API

Commit:
Merge remote-tracking branch 'origin/develop' into config-source

Conflicts:
	pom.xml

Commit:
revert API changes

Commit:
removed constructor

Commit:
remove not necessary changes

Commit:
cleanup pom and source

Commit:
removed unused imports

Commit:
bump to the next version of scalecube config

Commit:
Added JSON Object Mapper to the config source

Commit:
added javadocs

Commit:
added CR@EOF

artem-v commented on config-scalecube-configuration-service/pom.xml:
```diff
@@ -0,0 +1,85 @@
+      <exclusions>
+        <exclusion>
+          <groupId>io.netty</groupId>
+          <artifactId>netty-buffer</artifactId>
```
Do we still need this exclusion after new `scalecube-services` release?

artem-v commented on config-scalecube-configuration-service/pom.xml:
```diff
@@ -0,0 +1,85 @@
+      <exclusions>
+        <exclusion>
+          <groupId>io.rsocket</groupId>
+          <artifactId>rsocket-core</artifactId>
```
Maybe add another exclusion to exclude websocket gatewaqy as well?

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+          .collectMap(FetchResponse::key, this.parsing::fromFetchResponse)
+          .block();
+    } catch (Exception e) {
+      e.printStackTrace();
```
Remove

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+  private static class Parsing {
+    private ObjectWriter writer;
+
+    protected Parsing() {
```
You meant `private` maybe ?

artem-v commented on config-scalecube-configuration-service/src/test/java/io/scalecube/config/service/BrokerData.java:
```diff
@@ -0,0 +1,57 @@
+
+public class BrokerData {
+
+  private String BrokerID;
```
Maybe `brokerId` ?

aharonha replied:
This is the data structure in production

artem-v commented on pom.xml:
```diff
@@ -23,7 +24,7 @@
   <properties>
-    <scalecube.config.version>0.3.9</scalecube.config.version>
+    <scalecube.config.version>0.3.11</scalecube.config.version>
     <scalecube-services.version>2.5.10</scalecube-services.version>
```
scalecube-service `2.5.10` is a newest?

aharonha replied:
```suggestion
    <scalecube-services.version>2.6.0</scalecube-services.version>
```

artem-v commented on config-scalecube-configuration-service/src/test/java/io/scalecube/config/service/BrokerData.java:
```diff
@@ -0,0 +1,57 @@
+public class BrokerData {
+
+  private String BrokerID;
+  private ApiKey[] APIKeys;
```
Maybe `apiKeys` ?

aharonha replied:
This is the data structure in production

artem-v commented on config-scalecube-configuration-service/src/test/java/io/scalecube/config/service/ApiKey.java:
```diff
@@ -0,0 +1,57 @@
+
+public class ApiKey {
+
+  private String APIKey;
```
Maybe `apiKey` ?

aharonha replied:
this is the type name in the real production data.

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(ScalecubeConfigurationServiceConfigSource.class);
+
+  ConfigurationService service;
```
Put a `private final` please

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ObjectMapperHolder.java:
```diff
@@ -0,0 +1,77 @@
+      try {
+        return objectMapper.readerFor(toType).readValue(src);
+      } catch (IOException exception) {
+        ThrowableUtil.propagate(exception);
```
`throw`  and hence you can omit `return`ing

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+  public ScalecubeConfigurationServiceConfigSource(
+      String token, String repository, ConfigurationService service) {
+    this.service = service;
+    requestEntries = new EntriesRequest(token, repository);
```
Maybe just put `EntriesRequest` as a parameter instead of having two: `token`, and `repository`.  This you would decouple from signature constructor of `EntriesRequest` .

aharonha replied:
It makes no sense to do this. it's like exposing the way this config-source is working (and hereafter locking the implementation)

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+
+  ConfigurationService service;
+
+  private EntriesRequest requestEntries;
```
Put a `final` plz

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+    try {
+      return new URL("https", "configuration-service-http.genesis.om2.com", 443, "/");
+    } catch (MalformedURLException ignoredException) {
+      return null;
```
I suppose we have to throw  here

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+
+  private static URL defaultUrl() {
+    try {
+      return new URL("https", "configuration-service-http.genesis.om2.com", 443, "/");
```
Why it's hardcoded? Let's some component (runner) living at the edge  will decide what's _default_  for him.

aharonha replied:
The default is just a default. there is another constructor with URL.

artem-v commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+    this.service = http(builder.build()).forService(ConfigurationService.class);
+    requestEntries = new EntriesRequest(token, repository);
+  }
+
```
Upon constructor `ScalecubeConfigurationServiceConfigSource(String token, String repository, URL service)`.  IMHO this should be a  factory-method thing. IMHO constructor should be more lightweiht  and factory-method can be thick.

artem-v committed:
Merge branch 'develop' into config-source

segabriel commented on config-scalecube-configuration-service/pom.xml:
```diff
@@ -0,0 +1,85 @@
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-all</artifactId>
+      <version>1.9.5</version>
```
maybe it will better use version 2.x

aharonha replied:
version 2 is still beta now

segabriel commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ObjectMapperHolder.java:
```diff
@@ -0,0 +1,77 @@
+        return objectMapper.readerFor(toType).readValue(src);
+      } catch (IOException exception) {
+        ThrowableUtil.propagate(exception);
+        return null;
```
it's a utility class so it will better throw an exception (sure which will extend RuntimeException) instead `null`, because `null` value it's a valid value as well. And we can make the assumption that property was deleted but in actually it was an incorrect value.

segabriel commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+            .port(service.getPort())
+            .contentType(JacksonCodec.CONTENT_TYPE)
+            .loopResources(HttpResources.get());
+    if ("https".equals(service.getProtocol()) || "wss".equals(service.getProtocol())) {
```
so we check "https" as well as "wss" but only HTTP client will be created. Is it correct?

aharonha replied:
the wss is just being safe. if someone will use ftp://127.0.0.1:80 I will create it as non-secure. If you think that this is redundant I will remove it

segabriel replied:
no, I just don't understand why we create only HttpClient, we also can create WSClient and RsokcetClient.

segabriel commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+   * @param service the actual configuration service
+   */
+  public ScalecubeConfigurationServiceConfigSource(
+      String token, String repository, ConfigurationService service) {
```
Do we really need this constructor? Do we have a case to use it in such a manner?

segabriel commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+                fetchResponse.key(), writer.writeValueAsString(fetchResponse.value()))
+            .build();
+      } catch (JsonProcessingException ignoredException) {
+        return LoadedConfigProperty.withNameAndValue(fetchResponse.key(), null).build();
```
I think it's a bad idea to consider it as a null value due actually it's not that.

aharonha replied:
it was your idea in the first place

segabriel replied:
perhaps, but now, in the changed code I think we shouldn't create `LoadedConfigProperty` with the null value in error case. Just [don't update]/[skip] this value.

aharonha replied:
we should comply to the behavior of all others.

segabriel commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,124 @@
+    public ConfigProperty fromFetchResponse(FetchResponse fetchResponse) {
+      try {
+        return LoadedConfigProperty.withNameAndValue(
+                fetchResponse.key(), writer.writeValueAsString(fetchResponse.value()))
```
So just to understand we retrieve the `FetchResponse` and the Jackson turns value into `HashMap` structure, then we turn back this `HashMap` into `String` to put in the ConfigSource. Is that really?

aharonha replied:
yes

segabriel commented on config-scalecube-configuration-service/src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,18 @@
+  </Appenders>
+
+  <Loggers>
+    <Root level="DEBUG">
```
I think we can encounter with the Travis problem due to this chosen level

aharonha replied:
If we will get to this point we will fix it. up until now it's ok.

segabriel commented on config-scalecube-configuration-service/src/test/java/io/scalecube/config/service/ApiKey.java:
```diff
@@ -0,0 +1,57 @@
+    return result;
+  }
+
+  /* (non-Javadoc)
```
really informative java docs :+1:

aharonha replied:
this kind of javadocs is created automatically in order to make a link to the original one transparently.

segabriel replied:
so maybe let's change in your idea generation of these redundant javadocs or just remove always it yourself.

aharonha replied:
this is a simple POJO class to demonstrate in tests the abilities of this project.

Commit:
CR updates

Commit:
using a builder to create ScalecubeConfigurationServiceConfigSource

aharonha committed:
Update scalecube-services.version

dmytro-lazebnyi commented on config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java:
```diff
@@ -0,0 +1,133 @@
+
+  private static URL defaultUrl() {
+    try {
+      return new URL("https", "configuration-service-http.genesis.om2.com", 443, "/");
```
Let's do not hardcode anything that related to specific (internal) deployment infrastructure into an open source project.

Commit:
removed exclusions

Commit:
removed default url

and can build from either URL or a service

Commit:
CR: removed stupid printout

Commit:
CR private method parsing

Commit:
returning null when the JSON is broken

Commit:
+ javadocs

Commit:
Merge remote-tracking branch 'origin/develop' into config-source

Commit:
fix http & https protocol

Commit:
remove non javadocs comments

Commit:
update versions:

scalecube-services-gateway.version => 2.6.0
scalecube-services.version => 2.6.0

artem-v committed:
Fixed checkstyle and enforcer

artem-v committed:
Merge remote-tracking branch 'origin/config-source' into config-source

# Conflicts:
#	config-scalecube-configuration-service/src/main/java/io/scalecube/config/service/ScalecubeConfigurationServiceConfigSource.java

artem-v committed:
Fixed checkstyle and enforcer

Commit:
Merge branch 'config-source' of git@github.com:scalecube/scalecube-configuration-service.git into config-source

Commit:
Added example

aharonha committed:
Update Person.java

EOL@EOF

artem-v reviewed (changes requested):
Need to update on 2.7.0 gateway

artem-v committed:
Added gateway 2.7.0 dependency instead of old ones

Commit:
fix example pojo

Commit:
Merge branch 'config-source' of git@github.com:scalecube/scalecube-configuration-service.git into config-source
